ruma-appservice-api/src/event/push_events/v1.rs

```diff
@@ -36,6 +44,84 @@ impl<'a> Request<'a> {
     }
 }
 
+#[derive(Debug, Deserialize)]
+struct EventDeHelper {
+    /// the Matrix event type string "m.room.whatever".
+    #[serde(rename = "type")]
+    pub ev_type: String,
+
+    /// If `state_key` is present the event will be deserialized as a state event.
+    pub state_key: Option<IgnoredAny>,
+
+    /// If no `state_key` is found but an `event_id` is present the event
+    /// will be deserialized as a message event.
+    pub event_id: Option<IgnoredAny>,
+
+    /// If no `event_id` or `state_key` are found but a `room_id` is present
+    /// the event will be deserialized as a ephemeral event.
+    pub room_id: Option<RoomId>,
+
+    /// If this `UnsignedData` contains a `redacted_because` key the event is
+    /// immediately deserialized as a redacted event.
+    pub unsigned: Option<UnsignedDeHelper>,
+}
+
+// all observed room events go into "Joined Room" since appservice transactions
+// aren't bound to a specific user
+impl TryInto<sync_events::Response> for IncomingRequest {
+    type Error = serde_json::Error;
+
+    fn try_into(self) -> Result<sync_events::Response, Self::Error> {
+        let mut response = sync_events::Response::new(self.txn_id.clone());
+
+        for raw_event in self.events {
+            let helper: EventDeHelper = serde_json::from_str(raw_event.json().get())?;
+            let event_json = Raw::into_json(raw_event);
+
+            if let Some(room_id) = helper.room_id {
+                let join =
+                    response.rooms.join.entry(room_id).or_insert_with(sync_events::JoinedRoom::new);
+
+                if helper.state_key.is_some() {
+                    match helper.unsigned {
+                        Some(unsigned) if unsigned.redacted_because.is_some() => join
```

Comment:
Do redacted state events really go into `timeline` rather than `state`? That would be new to me.

Reply:
I'm not sure actually. I think I got the impression that it needs to be that way because there's no apparent way in the docs to go from `AnySyncStateEvent` to `AnyRedactedSyncStateEvent` in https://docs.rs/ruma/0.0.3/ruma/api/client/r0/sync/sync_events/struct.State.html, while there is an apparent way to go from `AnySyncRoomEvent` to `AnyRedactedSyncStateEvent` in https://docs.rs/ruma/0.0.3/ruma/api/client/r0/sync/sync_events/struct.Timeline.html?

Reply:
I guess I should even put the regular state event into the timeline as well. Will do some real-life synapse testing :D

Reply:
But reading the spec for /sync it says

> timeline  | The timeline of messages and state changes in the room.

So I suppose I need to put state and redacted state events into `state` and `timeline`

Reply:
Oh, right, state is only for state updates, and timeline has both. Now that I'm understanding this whole thing better, I wonder whether it can really be made to work. You're also using the `txn_id` above as the next_batch token (the argument to `sync_events::Response::new`, are these documented as correlating at all?

Reply:
> You're also using the txn_id above as the next_batch token (the argument to sync_events::Response::new, are these documented as correlating at all?

No, they don't correlate. That's why I originally said that I think it probably doesn't belong into ruma, because using `txn_id` as `next_batch` token can only be handled correctly if the crate consumer knows what they're doing. In the case of matrix-rust-sdk that simply means that we probably don't need the `next_batch` token because all events are pushed to us, so we never actively call `/sync` (or any API that requires the `batch_token`)

Reply:
In that case, can you make `next_batch` a parameter of the conversion mathod (as part of making it inherent rather than a `TryInto` impl)?

Reply:
Apart from that, when converting from Appservice transaction it probably makes sense to never put something into `state` but instead always in `timeline` because we never have state gaps thanks to the transactions

Reply:
> In that case, can you make next_batch a parameter of the conversion mathod (as part of making it inherent rather than a TryInto impl)?

Sounds good